### DISCUSSION TITLE
How can I host foocus in Hugging Spaces?

### DISCUSSION THREAD

**Discussion by hello-rangel** [lllyasviel/Fooocus#2512] [Q&A] [answered]
When I run 
```
!pip install pygit2==1.12.2
%cd /content
!git clone https://github.com/lllyasviel/Fooocus.git
%cd /content/Fooocus
!python entry_with_update.py --share

```
in drive collab I get this message: "This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from Terminal to deploy to Spaces (https://huggingface.co/spaces)"

could someone please guide me how I could do this? I created and account and tried to do it myself but I haven't find the way

**Comment by AndrewMTA5** [comment]
Trying to figure this out too

**Comment by mashb1t** [comment, marked as answer]
Here are a few words to HF Spaces:

Huggingface Spaces doesn't allow downloading of files, which is why you have to commit / build everything, including models, LoRAs, VAEs, and many more. 
Fooocus is built in the way that it downloads files on demand, which means when they're first loaded and do not exist. 
This is contrary to the approach HF Spaces follows. In order to run Fooocus in HF Spaces you need to upload all these files, roughly 15GB, or you edit the `README.md` file, see below.

If you don't, you'll receive this error:

```
Runtime error
end(msg)
  File "/usr/local/lib/python3.10/http/client.py", line 976, in send
    self.connect()
  File "/usr/local/lib/python3.10/http/client.py", line 1448, in connect
    super().connect()
  File "/usr/local/lib/python3.10/http/client.py", line 942, in connect
    self.sock = self._create_connection(
  File "/usr/local/lib/python3.10/socket.py", line 824, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/local/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Temporary failure in name resolution
```

Make sure you remove these lines from the .gitignore file:

```
*.ckpt
*.safetensors
*.pth
*.pt
*.bin
*.patch
*.backup
*.corrupted
*.partial
*.onnx
```

If you have previously run `gradio deploy` you have to clear the docker build cache in your Huggingface space.

Guide:
1. copy requirements_versions.txt to requirements.txt
2. run `gradio deploy`
3. add token
4. choose app title, e.g. `fooocus`
5. use app file `launch.py`
6. use hardware `t4-small` as the free CPU instance cpu-basic will take ages to generate images
7. no secrets needed
8. no actions needed

This creates a file called `README.md` with the HF config.
After this the command will upload the ~15GB of files, if you've set up Fooocus locally. Alternatively, you can also configure each and every model in `preload_from_hub`, see reference https://huggingface.co/docs/hub/spaces-config-reference

`app_port` also should be set to 7865.

This also marks the step where i have stopped testing and why HF Spaces isn't supported by default in Fooocus. 
It may be easier to build a docker container including the download of all models with `ADD <URL>`.

HF Spaces also doesn't offer a persistent file system by default (only paid), so you might have to add `--disable-image-log` (haven't tested, also don't know where exactly to add it., I'd guess after `app_file: launch.py --disable-image-log`, but please check yourself. It might also not even work.).

Have Fun with Spaces, but maybe consider using another service if you aim for ease-of-use, like [Colab](https://colab.research.google.com/github/lllyasviel/Fooocus/blob/main/fooocus_colab.ipynb) (Free/Pro), [Rundiffusion](https://rundiffusion.com/) or in general any other Linux server hoster.

**Reply by hello-rangel** [reply]
thank you so much for taking the time for this detail response!

**Reply by hello-rangel** [reply]
I have been scrapping the web for some time now, could you suggest a Linux server?

**Reply by mashb1t** [reply]
Depends on your use-case, image volume and price range as well as if you plan to work with it also long-term. Cloud solutions are in general best if you want to either share access to other users or only use it short to medium term. Can you provide more insights in your intended use?

**Reply by hello-rangel** [reply]
Pretty much foocus is too slow on my PC and sometimes it freezes the browser. With Collab, the free version is limited and multiple times tell me that it has used all the RAM. I just want to run foocus, I am planning on using it long-term for my own use. Open to multiple solutions

**Reply by mashb1t** [reply]
to be honest i'd then recommend to get a personal rig with a good graphics card instead of a cloud server, as it's cheaper and more performant in the long term.vDepending on your already existing system this may also not even cost that much, only PSU and graphics card upgrade might be needed. You can get a good GPU for around 600+$
If you only want Fooocus on demand you can also go for Colab Pro, but let's say you want a server with A10 GPU (~3x faster than T4) you'll have to pay around 800-900$ per month, which is on demand price and no savings plan.
If you eant to go for a T4 GPU it will be around 380$/month.
Running Fooocus is still cheaper if you only need it on demand, even Colab Free is perfectly fine. The drawback is that you get eisconnected from the server every 72h, but that may be acceptible. Also upgrading to Colab Pro could help.
I hope this gave you some insights and provides you a better base for your decision.

**Reply by gsw85** [reply]
If I want to offer fooocus as a service, built my own gui and send data to fooocus (prompt , img , masked img) , which one should go for?

**Reply by mashb1t** [reply]
@gsw85 that depends on multiple things:
- cloud provider you're most familiar with
- auto-scaling capabilities and auto-recovering / failover
- regiona and availability zones and availability for instances in those zones
- cost for instances
- amount of users
and many more

**Reply by gsw85** [reply]
@mashb1t , I am not really familiar with any particular cloud service provider, the cost of instance would be go for minimum required, 

I am thinking of using replicate.com / huggingface space / runpod probably? - that would no need to worry about scaling?

**Reply by mashb1t** [reply]
@gsw85 you need min. T4, but also then you're restricted and can only use a limited amount of features in Fooocus (similar to Colab, see docs [here](https://github.com/lllyasviel/Fooocus?tab=readme-ov-file#colab)).

> Note that this Colab will disable refiner by default because Colab free's resources are relatively limited (and some "big" features like image prompt may cause free-tier Colab to disconnect). We make sure that basic text-to-image is always working on free-tier Colab.
Using --always-high-vram shifts resource allocation from RAM to VRAM and achieves the overall best balance between performance, flexibility and stability on the default T4 instance. Please find more information [here](https://github.com/lllyasviel/Fooocus/pull/1710#issuecomment-1989185346).

Just fyi for huggingface spaces you need to select an instance or be subscribed to [HF PRO](https://huggingface.co/subscribe/pro) / [HF Enterprise](https://huggingface.co/enterprise) to get access to [ZeroGPU Spaces](https://huggingface.co/new-space?sdk=gradio)

**Reply by gsw85** [reply]
@mashb1t  Alright thanks, I will try on the HF T4
